### PR TITLE
feat(lang-v2): support PDA payers for init constraints

### LINKED ISSUE #1120: Allow PDAs to be used as payers
relevant links
https://pastebin.com/QnJVhmex
https://gist.github.com/cqfd/ffbdebbbcd3460ec822c3e63572cc282
https://discord.com/channels/889577356681945098/889702325231427584/918625031209246751

tl;dr there would be different code for PDAs owned by the deriving program vs PDAs owned by the system program

willing to fund this issue or any other issues ive made

### PR DESCRIPTION
**Pull request by Otter-0x4ka5h**
Closes #1120

### CODE REVIEW

File: lang-v2/derive/src/parse.rs [hunk 1]
```diff
@@ -961,13 +971,125 @@ fn emit_seeds_check(
...
+            let __payer_signer_seeds: Option<&[&[u8]]> =
+                Some(&[#(#seed_refs),* , __payer_bump_seed.as_ref()]);
+        });
+    }
+
+    if let Some(Some(ref bump_expr)) = payer_field.attrs.bump {
+        return Ok(quote! {
+            let __payer_seed_expr_val = #seeds_expr;
+            let __payer_seed_ref: &[&[u8]] = __payer_seed_expr_val.as_ref();
+            if __payer_seed_ref.len() > 16 {
```
Comment (swaroop-osec) [inline lang-v2/derive/src/parse.rs]: Nit: `16` and `17` show up as bare literals in `emit_payer_signer_seeds_binding` they're Solana's `MAX_SEEDS` and `MAX_SEEDS + 1` (for the appended bump byte). The intent isn't obvious on a first read. It's better lifting into pub consts on `anchor_lang_v2` so the emitted code reads `MAX_PAYER_SEEDS` / `MAX_PAYER_SEEDS_WITH_BUMP` instead of bare numbers easier to grasp the intent. Cosmetic, not blocking.


```rust
// in lang-v2/src/lib.rs
pub const MAX_PAYER_SEEDS: usize = 16;
pub const MAX_PAYER_SEEDS_WITH_BUMP: usize = MAX_PAYER_SEEDS + 1;
```

File: tests-v2/tests/pda_payer.rs
Comment (swaroop-osec) [inline tests-v2/tests/pda_payer.rs:1]: Can we increase test coverage for edge cases Option<SystemAccount> as payer, Box<Account<T>> initialized with a PDA payer, and a PDA payer with more than 16 seeds (should fail)?

**Review by swaroop-osec — DISMISSED** [review]
Can you address the changes?

File: tests-v2/tests/pda_payer.rs
Comment (0x4ka5h) [reply tests-v2/tests/pda_payer.rs:1]: added

File: lang-v2/derive/src/parse.rs (hunk 1 above)
Comment (0x4ka5h) [reply lang-v2/derive/src/parse.rs]: address in the latest commit

**Review by swaroop-osec — APPROVED** [review]
lgtm

File: lang-v2/derive/src/parse.rs [hunk 2]
```diff
@@ -961,13 +971,134 @@ fn emit_seeds_check(
...
+        let __payer_bump =
+            anchor_lang_v2::find_and_verify_program_address(
+                __payer_seed_ref, __program_id, __payer.address(),
+            ).map_err(|_| anchor_lang_v2::ErrorCode::ConstraintSeeds)?;
+        __bumps.#bump_field = __payer_bump;
+        let mut __payer_seed_buf: [&[u8]; anchor_lang_v2::MAX_PAYER_SEEDS_WITH_BUMP] =
+            [&[]; anchor_lang_v2::MAX_PAYER_SEEDS_WITH_BUMP];
+        let __payer_seed_count = __payer_seed_ref.len();
+        __payer_seed_buf[..__payer_seed_count].copy_from_slice(__payer_seed_ref);
+        __payer_seed_buf[__payer_seed_count] = &[__payer_bump];
```
Comment (jamie-osec) [inline lang-v2/derive/src/parse.rs]: This causes a 'temporary value dropped while borrowed' error; please add the `let __payer_bump_bytes = [__payer_bump];` temporary used in the explicit-bump branch and ensure this case is exercised:

```rs
#[derive(Accounts)]
pub struct OpaquePayerSeeds {
    #[account(mut, seeds = payer_seeds(), bump)]
    pub payer: SystemAccount,
    #[account(init, payer = payer, space = 8 + core::mem::size_of::<Data>())]
    pub data: Account<Data>,
    pub system_program: Program<System>,
}
```

**Review by jamie-osec — APPROVED** [review]
